File: backend-node/routes/inventory.js
const express = require('express');
const router = express.Router();
const { Op, fn, col, literal } = require('sequelize');
const { Material, InventoryItem, StockMovement, ReorderAlert, sequelize } = require('../models');
const { authenticateToken } = require('../middleware/auth');

// 1. List all materials
router.get('/materials', authenticateToken, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 100;
    const offset = parseInt(req.query.skip) || 0;
    const { category, search } = req.query;

    const whereClause = { is_active: true };

    if (category) {
      whereClause.category = category;
    }

    if (search) {
      whereClause[Op.or] = [
        { name: { [Op.like]: `%${search}%` } },
        { material_code: { [Op.like]: `%${search}%` } }
      ];
    }

    const materials = await Material.findAll({
      where: whereClause,
      limit,
      offset
    });

    return res.json(materials);
  } catch (error) {
    console.error('List Materials Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 2. Add new material
router.post('/materials', authenticateToken, async (req, res) => {
  try {
    const { material_code, name, category, unit, reorder_level, reorder_quantity, unit_cost } = req.body;

    const existing = await Material.findOne({ where: { material_code } });
    if (existing) {
      return res.status(400).json({ detail: 'Material code already exists' });
    }

    const newMaterial = await Material.create({
      material_code,
      name,
      category,
      unit,
      reorder_level,
      reorder_quantity,
      unit_cost
    });

    return res.status(201).json(newMaterial);
  } catch (error) {
    console.error('Create Material Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 3. Get material details with current stock
router.get('/materials/:material_id', authenticateToken, async (req, res) => {
  try {
    const material_id = parseInt(req.params.material_id);
    const material = await Material.findByPk(material_id);
    if (!material) {
      return res.status(404).json({ detail: 'Material not found' });
    }

    const current_stock_val = await InventoryItem.sum('quantity', {
      where: { material_id }
    });

    return res.json({
      material,
      current_stock: parseFloat(current_stock_val || 0)
    });
  } catch (error) {
    console.error('Get Material Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 4. Get current stock levels
router.get('/stock', authenticateToken, async (req, res) => {
  try {
    const { material_id, location, low_stock } = req.query;

    const whereClause = {};
    const itemWhereClause = {};

    if (material_id) {
      whereClause.material_id = parseInt(material_id);
    }
    if (location) {
      itemWhereClause.location = location;
    }

    // Since we need grouping, we can fetch materials with inventory items
    // and sum in Javascript or use raw SQL/Sequelize attributes.
    // Let's use a query to aggregate.
    const results = await Material.findAll({
      where: whereClause,
      attributes: [
        'material_id',
        'material_code',
        'name',
        'unit',
        'reorder_level',
        [fn('COALESCE', fn('SUM', col('inventory_items.quantity')), 0), 'current_stock'],
        [fn('COALESCE', fn('SUM', col('inventory_items.total_value')), 0), 'stock_value']
      ],
      include: [{
        model: InventoryItem,
        as: 'inventory_items',
        attributes: [],
        where: Object.keys(itemWhereClause).length ? itemWhereClause : undefined,
        required: false // LEFT OUTER JOIN to include materials with 0 stock
      }],
      group: ['Material.material_id'],
      raw: true
    });

    const stock_data = [];
    for (const row of results) {
      const current_stock = parseFloat(row.current_stock || 0);
      const reorder_level = parseFloat(row.reorder_level || 0);
      const stock_value = parseFloat(row.stock_value || 0);
      const is_low_stock = current_stock < reorder_level;

      const stock_info = {
        material_id: row.material_id,
        material_code: row.material_code,
        name: row.name,
        unit: row.unit,
        current_stock,
        reorder_level,
        stock_value,
        is_low_stock
      };

      if (low_stock === 'true' || low_stock === true) {
        if (is_low_stock) stock_data.push(stock_info);
      } else {
        stock_data.push(stock_info);
      }
    }

    return res.json({
      total_items: stock_data.length,
      stock: stock_data
    });

  } catch (error) {
    console.error('Get Stock Levels Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 5. Receive stock into inventory
router.post('/stock/receive', authenticateToken, async (req, res) => {
  const t = await sequelize.transaction();
  try {
    const { material_id, quantity, batch_number, location, po_id, unit_cost } = req.body;

    const material = await Material.findByPk(material_id);
    if (!material) {
      await t.rollback();
      return res.status(404).json({ detail: 'Material not found' });
    }

    const batch = batch_number || `BATCH-${Date.now()}`;
    const cost = unit_cost || material.unit_cost || 0;
    const total_value = quantity * cost;

    const inventory_item = await InventoryItem.create({
      material_id,
      batch_number: batch,
      quantity,
      location: location || 'MAIN-WAREHOUSE',
      unit_cost: cost,
      total_value,
      po_id,
      quality_status: 'approved'
    }, { transaction: t });

    await StockMovement.create({
      material_id,
      movement_type: 'receipt',
      quantity,
      batch_number: batch,
      to_location: inventory_item.location,
      reference_type: po_id ? 'purchase_order' : 'manual',
      reference_id: po_id,
      unit_cost: cost,
      total_value,
      created_by: req.user.user_id
    }, { transaction: t });

    await t.commit();

    return res.json({
      message: 'Stock received successfully',
      batch_number: batch,
      quantity: parseFloat(quantity)
    });

  } catch (error) {
    await t.rollback();
    console.error('Receive Stock Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 6. Issue stock from inventory
router.post('/stock/issue', authenticateToken, async (req, res) => {
  const t = await sequelize.transaction();
  try {
    const { material_id, quantity, batch_number, wo_id, notes } = req.body;

    let inventory_item;

    if (batch_number) {
      inventory_item = await InventoryItem.findOne({
        where: { material_id, batch_number },
        transaction: t
      });
    } else {
      inventory_item = await InventoryItem.findOne({
        where: { material_id, quantity: { [Op.gt]: 0 } },
        transaction: t
      });
    }

    if (!inventory_item || parseFloat(inventory_item.quantity) < quantity) {
      await t.rollback();
      return res.status(400).json({ detail: 'Insufficient stock' });
    }

    // Deduct quantity
    const newQuantity = parseFloat(inventory_item.quantity) - parseFloat(quantity);
    inventory_item.quantity = newQuantity;
    inventory_item.total_value = newQuantity * parseFloat(inventory_item.unit_cost);
    await inventory_item.save({ transaction: t });

    // Create stock movement
    await StockMovement.create({
      material_id,
      movement_type: 'issue',
      quantity,
      batch_number: inventory_item.batch_number,
      from_location: inventory_item.location,
      reference_type: wo_id ? 'work_order' : 'manual',
      reference_id: wo_id,
      unit_cost: inventory_item.unit_cost,
      total_value: quantity * parseFloat(inventory_item.unit_cost),
      notes,
      created_by: req.user.user_id
    }, { transaction: t });

    await t.commit();

    return res.json({
      message: 'Stock issued successfully',
      batch_number: inventory_item.batch_number,
      quantity: parseFloat(quantity)
    });

  } catch (error) {
    await t.rollback();
    console.error('Issue Stock Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 7. Get stock movement history
router.get('/stock/movements', authenticateToken, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 100;
    const offset = parseInt(req.query.skip) || 0;
    const { material_id, movement_type } = req.query;

    const whereClause = {};
    if (material_id) {
      whereClause.material_id = parseInt(material_id);
    }
    if (movement_type) {
      whereClause.movement_type = movement_type;
    }

    const count = await StockMovement.count({ where: whereClause });
    const movements = await StockMovement.findAll({
      where: whereClause,
      order: [['created_at', 'DESC']],
      limit,
      offset
    });

    return res.json({
      total: count,
      movements
    });
  } catch (error) {
    console.error('Get Stock Movements Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 8. Get detailed stock ledger for a material
router.get('/stock/ledger', authenticateToken, async (req, res) => {
  try {
    const material_id = parseInt(req.query.material_id);
    const material = await Material.findByPk(material_id);
    if (!material) {
      return res.status(404).json({ detail: 'Material not found' });
    }

    const movements = await StockMovement.findAll({
      where: { material_id },
      order: [['created_at', 'ASC']]
    });

    const current_stock_val = await InventoryItem.sum('quantity', {
      where: { material_id }
    });

    return res.json({
      material,
      current_stock: parseFloat(current_stock_val || 0),
      movements
    });
  } catch (error) {
    console.error('Get Stock Ledger Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 9. Get reorder alerts for low stock items
router.get('/reorder-alerts', authenticateToken, async (req, res) => {
  try {
    const { priority } = req.query;

    // Fetch low stock materials
    // We can do this with raw SQL query or standard Sequelize. Since it requires group by and having, a literal is cleaner:
    const low_stock_materials = await Material.findAll({
      attributes: [
        'material_id',
        'material_code',
        'name',
        'reorder_level',
        'reorder_quantity',
        [fn('COALESCE', fn('SUM', col('inventory_items.quantity')), 0), 'current_stock']
      ],
      include: [{
        model: InventoryItem,
        as: 'inventory_items',
        attributes: [],
        required: false
      }],
      group: ['Material.material_id'],
      having: literal('current_stock < Material.reorder_level'),
      raw: true
    });

    let alerts = [];
    for (const row of low_stock_materials) {
      const current = parseFloat(row.current_stock || 0);
      const reorder = parseFloat(row.reorder_level || 0);

      let priority_level = 'low';
      if (current === 0) {
        priority_level = 'critical';
      } else if (current < reorder * 0.3) {
        priority_level = 'high';
      } else if (current < reorder * 0.6) {
        priority_level = 'medium';
      }

      alerts.append = {
        material_id: row.material_id,
        material_code: row.material_code,
        name: row.name,
        current_stock: current,
        reorder_level: reorder,
        recommended_quantity: parseFloat(row.reorder_quantity || reorder),
        priority: priority_level
      };
      
      alerts.push(alerts.append);
    }

    if (priority) {
      alerts = alerts.filter(a => a.priority === priority);
    }

    return res.json({
      total_alerts: alerts.length,
      alerts
    });

  } catch (error) {
    console.error('Get Reorder Alerts Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 10. Trigger automatic reorder (Phase 2 placeholder)
router.post('/reorder-alerts/auto', authenticateToken, async (req, res) => {
  return res.json({
    message: 'Auto-reorder feature will be available in Phase 2',
    status: 'pending'
  });
});

module.exports = router;

File: backend/app/api/routes/inventory.py
"""
Inventory API Routes
Materials, Stock Management, Movements
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func, or_
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.auth import User
from app.models.inventory import Material, InventoryItem, StockMovement, ReorderAlert, MovementType, MaterialCategory

router = APIRouter()

# Schemas
class MaterialCreate(BaseModel):
    material_code: str
    name: str
    category: str
    unit: str
    reorder_level: Optional[float] = None
    reorder_quantity: Optional[float] = None
    unit_cost: Optional[float] = None

class MaterialResponse(BaseModel):
    material_id: int
    material_code: str
    name: str
    category: str
    unit: str
    reorder_level: Optional[float]
    
    class Config:
        from_attributes = True

class StockReceive(BaseModel):
    material_id: int
    quantity: float
    batch_number: Optional[str] = None
    location: Optional[str] = None
    po_id: Optional[int] = None
    unit_cost: Optional[float] = None

class StockIssue(BaseModel):
    material_id: int
    quantity: float
    batch_number: Optional[str] = None
    wo_id: Optional[int] = None
    notes: Optional[str] = None

# Material Routes
@router.get("/materials", response_model=List[MaterialResponse])
async def list_materials(
    skip: int = 0,
    limit: int = 100,
    category: Optional[str] = None,
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all materials"""
    query = db.query(Material).filter(Material.is_active == True)
    
    if category:
        query = query.filter(Material.category == category)
    
    if search:
        query = query.filter(
            or_(
                Material.name.ilike(f"%{search}%"),
                Material.material_code.ilike(f"%{search}%")
            )
        )
    
    materials = query.offset(skip).limit(limit).all()
    return materials

@router.post("/materials", response_model=MaterialResponse, status_code=201)
async def create_material(
    material_data: MaterialCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Add new material"""
    if db.query(Material).filter(Material.material_code == material_data.material_code).first():
        raise HTTPException(status_code=400, detail="Material code already exists")
    
    new_material = Material(**material_data.dict())
    db.add(new_material)
    db.commit()
    db.refresh(new_material)
    
    return new_material

@router.get("/materials/{material_id}")
async def get_material(
    material_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get material details with current stock"""
    material = db.query(Material).filter(Material.material_id == material_id).first()
    if not material:
        raise HTTPException(status_code=404, detail="Material not found")
    
    # Get current stock
    current_stock = db.query(func.sum(InventoryItem.quantity)).filter(
        InventoryItem.material_id == material_id
    ).scalar() or 0
    
    return {
        "material": material,
        "current_stock": float(current_stock)
    }

# Stock Routes
@router.get("/stock")
async def get_stock_levels(
    material_id: Optional[int] = None,
    location: Optional[str] = None,
    low_stock: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get current stock levels"""
    # Query to get stock by material
    query = db.query(
        Material.material_id,
        Material.material_code,
        Material.name,
        Material.unit,
        Material.reorder_level,
        func.sum(InventoryItem.quantity).label('current_stock'),
        func.sum(InventoryItem.total_value).label('stock_value')
    ).join(InventoryItem).group_by(Material.material_id)
    
    if material_id:
        query = query.filter(Material.material_id == material_id)
    
    if location:
        query = query.filter(InventoryItem.location == location)
    
    results = query.all()
    
    stock_data = []
    for row in results:
        stock_info = {
            "material_id": row.material_id,
            "material_code": row.material_code,
            "name": row.name,
            "unit": row.unit,
            "current_stock": float(row.current_stock or 0),
            "reorder_level": float(row.reorder_level or 0),
            "stock_value": float(row.stock_value or 0),
            "is_low_stock": (row.current_stock or 0) < (row.reorder_level or 0)
        }
        
        if not low_stock or stock_info["is_low_stock"]:
            stock_data.append(stock_info)
    
    return {
        "total_items": len(stock_data),
        "stock": stock_data
    }

@router.post("/stock/receive")
async def receive_stock(
    receive_data: StockReceive,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Receive stock into inventory"""
    material = db.query(Material).filter(Material.material_id == receive_data.material_id).first()
    if not material:
        raise HTTPException(status_code=404, detail="Material not found")
    
    # Create inventory item
    inventory_item = InventoryItem(
        material_id=receive_data.material_id,
        batch_number=receive_data.batch_number or f"BATCH-{datetime.now().strftime('%Y%m%d%H%M%S')}",
        quantity=receive_data.quantity,
        location=receive_data.location or "MAIN-WAREHOUSE",
        unit_cost=receive_data.unit_cost or material.unit_cost,
        total_value=receive_data.quantity * (receive_data.unit_cost or material.unit_cost or 0),
        po_id=receive_data.po_id,
        quality_status="approved"
    )
    
    db.add(inventory_item)
    
    # Create stock movement record
    movement = StockMovement(
        material_id=receive_data.material_id,
        movement_type=MovementType.RECEIPT,
        quantity=receive_data.quantity,
        batch_number=inventory_item.batch_number,
        to_location=inventory_item.location,
        reference_type="purchase_order" if receive_data.po_id else "manual",
        reference_id=receive_data.po_id,
        unit_cost=inventory_item.unit_cost,
        total_value=inventory_item.total_value,
        created_by=current_user.user_id
    )
    
    db.add(movement)
    db.commit()
    
    return {
        "message": "Stock received successfully",
        "batch_number": inventory_item.batch_number,
        "quantity": float(receive_data.quantity)
    }

@router.post("/stock/issue")
async def issue_stock(
    issue_data: StockIssue,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Issue stock from inventory"""
    # Find inventory item
    if issue_data.batch_number:
        inventory_item = db.query(InventoryItem).filter(
            InventoryItem.material_id == issue_data.material_id,
            InventoryItem.batch_number == issue_data.batch_number
        ).first()
    else:
        inventory_item = db.query(InventoryItem).filter(
            InventoryItem.material_id == issue_data.material_id,
            InventoryItem.quantity > 0
        ).first()
    
    if not inventory_item:
        raise HTTPException(status_code=404, detail="Insufficient stock")
    
    if inventory_item.quantity < issue_data.quantity:
        raise HTTPException(status_code=400, detail=f"Insufficient quantity. Available: {inventory_item.quantity}")
    
    # Deduct quantity
    inventory_item.quantity -= issue_data.quantity
    inventory_item.total_value = inventory_item.quantity * inventory_item.unit_cost
    
    # Create stock movement
    movement = StockMovement(
        material_id=issue_data.material_id,
        movement_type=MovementType.ISSUE,
        quantity=issue_data.quantity,
        batch_number=inventory_item.batch_number,
        from_location=inventory_item.location,
        reference_type="work_order" if issue_data.wo_id else "manual",
        reference_id=issue_data.wo_id,
        unit_cost=inventory_item.unit_cost,
        total_value=issue_data.quantity * inventory_item.unit_cost,
        notes=issue_data.notes,
        created_by=current_user.user_id
    )
    
    db.add(movement)
    db.commit()
    
    return {
        "message": "Stock issued successfully",
        "batch_number": inventory_item.batch_number,
        "quantity": float(issue_data.quantity)
    }

@router.get("/stock/movements")
async def get_stock_movements(
    material_id: Optional[int] = None,
    movement_type: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get stock movement history"""
    query = db.query(StockMovement)
    
    if material_id:
        query = query.filter(StockMovement.material_id == material_id)
    if movement_type:
        query = query.filter(StockMovement.movement_type == movement_type)
    
    movements = query.order_by(StockMovement.created_at.desc()).offset(skip).limit(limit).all()
    
    return {
        "total": query.count(),
        "movements": movements
    }

@router.get("/stock/ledger")
async def get_stock_ledger(
    material_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get detailed stock ledger for a material"""
    material = db.query(Material).filter(Material.material_id == material_id).first()
    if not material:
        raise HTTPException(status_code=404, detail="Material not found")
    
    movements = db.query(StockMovement).filter(
        StockMovement.material_id == material_id
    ).order_by(StockMovement.created_at).all()
    
    current_stock = db.query(func.sum(InventoryItem.quantity)).filter(
        InventoryItem.material_id == material_id
    ).scalar() or 0
    
    return {
        "material": material,
        "current_stock": float(current_stock),
        "movements": movements
    }

@router.get("/reorder-alerts")
async def get_reorder_alerts(
    status: Optional[str] = None,
    priority: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get reorder alerts for low stock items"""
    # Check for materials below reorder level
    low_stock_materials = db.query(
        Material.material_id,
        Material.material_code,
        Material.name,
        Material.reorder_level,
        Material.reorder_quantity,
        func.sum(InventoryItem.quantity).label('current_stock')
    ).join(
        InventoryItem, Material.material_id == InventoryItem.material_id, isouter=True
    ).group_by(
        Material.material_id
    ).having(
        func.coalesce(func.sum(InventoryItem.quantity), 0) < Material.reorder_level
    ).all()
    
    alerts = []
    for row in low_stock_materials:
        current = float(row.current_stock or 0)
        reorder = float(row.reorder_level or 0)
        
        # Determine priority
        if current == 0:
            priority_level = "critical"
        elif current < reorder * 0.3:
            priority_level = "high"
        elif current < reorder * 0.6:
            priority_level = "medium"
        else:
            priority_level = "low"
        
        alerts.append({
            "material_id": row.material_id,
            "material_code": row.material_code,
            "name": row.name,
            "current_stock": current,
            "reorder_level": reorder,
            "recommended_quantity": float(row.reorder_quantity or reorder),
            "priority": priority_level
        })
    
    if priority:
        alerts = [a for a in alerts if a["priority"] == priority]
    
    return {
        "total_alerts": len(alerts),
        "alerts": alerts
    }

@router.post("/reorder-alerts/auto")
async def auto_reorder(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Trigger automatic reorder for low stock items (Phase 2)"""
    # This will be implemented in Phase 2 with automation scripts
    return {
        "message": "Auto-reorder feature will be available in Phase 2",
        "status": "pending"
    }
